db/commitlog/commitlog.cc

```diff
@@ -1230,6 +1278,119 @@ db::commitlog::segment_manager::segment_manager(config c)
+    auto nt = _message_queue.front().timeout;
+    if (_message_loop_timer.armed() && nt < _message_loop_timer.get_timeout()) {
+        auto now = message_clock::now();
+        if (nt < now) {
+            _message_loop_timer.cancel();
+            wakeup();
+        } else {
+            _message_loop_timer.rearm(nt);
+        }
+    }
+}
+
+void db::commitlog::segment_manager::wakeup() {
+    if (_message_loop_wakeup) {
+        _message_loop_wakeup->set_value();
+        _message_loop_wakeup = nullptr;
+    }
+    ++_message_loop_wakeups;
+}
+
+struct db::commitlog::segment_manager::message_wait {
+    segment_manager& _manager;
+    promise<> _promise;
+
+    message_wait(segment_manager& m)
+        : _manager(m)
+    {
+        assert(_manager._message_loop_wakeup == nullptr);
+        _manager._message_loop_wakeup = &_promise;
+    }
+    ~message_wait() {
+        _manager._message_loop_wakeup = nullptr;
+    }
+    future<> get_future() {
+        return _promise.get_future();
+    }
+};
+
+class db::commitlog::segment_manager::message_partition {
+private:
+    std::vector<message>& _messages;
+    typename std::vector<message>::iterator _iter;
+public:
+    message_partition(std::vector<message>& messages, size_t index)
+        : _messages(messages)
+        , _iter(std::partition(messages.begin(), messages.end(), [index](auto& m) { return m.payload.index() != index; }))
+    {}
+    template<typename Cmp>
+    message_partition(std::vector<message>& messages, size_t index, Cmp&& cmp)
+        : message_partition(messages, index)
+    {
+        std::sort(_iter, messages.end(), cmp);
+    }
+    ~message_partition() {
+        _messages.erase(_iter, _messages.end());
+    }
+    auto begin() const {
+        return _iter;
+    }
+    auto end() const {
+        return _messages.end();
+    }
+    bool empty() const {
+        return _iter == _messages.end();
+    }
+};
+
+future<> db::commitlog::segment_manager::message_loop() {
+    std::vector<message> messages;
+    messages.reserve(10);
+
+    auto get_messages = [&](auto timeout) {
+        while (!_message_queue.empty() && _message_queue.front().timeout <= timeout) {
+            std::pop_heap(_message_queue.begin(), _message_queue.end(), mcmp);
+            messages.emplace_back(std::move(_message_queue.back()));
+            _message_queue.pop_back();
+        }
+    };
+
+    future<> tasks = make_ready_future<>();
+    bool shutdown = false;
+
+    while (!shutdown || !_message_queue.empty()) {
+        auto next_wakeup = no_timeout;
+        auto now = message_clock::now();
+        get_messages(shutdown ? no_timeout : now);
+
+        if (messages.empty() && !std::exchange(_message_loop_wakeups, 0u)) {
+            message_wait mw(*this);
+            _message_loop_timer.arm(next_wakeup);
+            co_await mw.get_future();
+            _message_loop_timer.cancel();
+            get_messages(message_clock::now());
+        }
```

Comment:
I totally like the idea of a message loop, but this is implementing a mini-Seastar on top of Seastar. Even the priority queue of the timer expirations. I'm sure there's a simpler way.

Reply:
As stated in the PR comment, the time-prio thing is an overkill. So far, we are only using the raw "wake up now" functionality. Otoh, the code to implement "wait until signalled or expired" is not hard, and we do need the expiration anyway to deal with the periodic aspect of the loop.